src/middlewares/Cache.ts

```diff
@@ -0,0 +1,155 @@
+    metadata?: {[index: string]: any};
+}
+
+// Represents cache options for each cache buckets added to the cache middleware
+export interface IFetchRequestCacheOptions {
+    enable: boolean;
+    path: string; // relative path, shall be resolved from FetchRequest.baseUrl
+    maxAge?: number; // cache ttl in seconds
+    key?: string; // optionally a specific key to be used in place of the generated hash
+}
+
+// Represents a signature of cache bucket, each bucket can respond to certain endpoint, have its own cache key and ttl
+export interface ICacheBucket {
+    options: IFetchRequestCacheOptions;
+}
+
+// Represents the storage engine functionality
+export interface ICacheMiddlewareStore {
+    setItem(key: string, value: string, maxAge: number): void;
+    getItem(key: string): string|null;
+    removeItem(key: string): void;
+}
+
+// Represents the cache middleware global configration
+export interface ICacheConfiguration {
+    maxAge?: number;
+    storage?: string;
+    namespace?: string; // prepends cache key useful when using multiple service clients &/or cache providers
+}
+
+// noinspection TsLint
+declare const Response: FetchResponse<string>;
+
+export class Cache implements IMiddleware<FetchRequest, Promise<FetchResponse<any>>> {
+    protected buckets: ICacheBucket[] = [];
+    protected storage: ICacheMiddlewareStore;
+
+    /**
+     * Constructor.
+     *
+     * Accepts global configuration, sets default storage engine.
+     *
+     * @param {ICacheConfiguration} config
+     */
+    public constructor(protected config: ICacheConfiguration = {}) {
+        this.storage = this.getDefaultCacheStorage();
+    }
+
+    /**
+     * Adds a new cache bucket to the stack. Each bucket can respond to certain endpoint, have its own cache key and ttl.
+     *
+     * @param {IFetchRequestCacheOptions} options
+     */
+    public addBucket(options: IFetchRequestCacheOptions): void {
+        this.buckets.push({ options });
+    }
+
+    /**
+     * Process the request. First try if we have cache and serve right away,
+     * else let the next middleware in pileline be invoked and cache it.
+     *
+     * @param  {FetchRequest} options
+     * @param  {callback}       next
+     * @return {any}
+     */
+    public process(options: FetchRequest, next: (nextOptions: FetchRequest) => Promise<FetchResponse<any>>): any {
+        const bucket = this.getEffectiveBucket(options);
+        const cacheable: boolean = (options.method || "GET").toLowerCase() === "get";
+
+        return cacheable && bucket && this.getCachedResponse(options, bucket)
+        || next(options).then((response: FetchResponse<any>) => {
+            if (!cacheable || !bucket) { return response; }
+
+            const clone = response.clone();
+            const ttl = bucket.options.maxAge || this.config.maxAge || 600; // TTL fallback to 10 min
+            clone.text().then((text: string) => this.storage.setItem(this.getCacheKey(options, bucket), text, ttl) );
```

Comment:
the use of `expiry` instead of `create timestamp` is for each of usage during retrieval, without needing to do the maths everytime.
for now the text dump of response nicely can mirror original response for consumption. i do think one thing is lacking though, we should attach http headers too.